sklearn/metrics/_plot/confusion_matrix.py

```diff
@@ -83,22 +83,30 @@ def plot(self, include_values=True, cmap='viridis',
         n_classes = cm.shape[0]
         self.im_ = ax.imshow(cm, interpolation='nearest', cmap=cmap)
         self.text_ = None
-
         cmap_min, cmap_max = self.im_.cmap(0), self.im_.cmap(256)
 
         if include_values:
             self.text_ = np.empty_like(cm, dtype=object)
-            if values_format is None:
-                values_format = '.2g'
 
             # print text with appropriate color depending on background
             thresh = (cm.max() + cm.min()) / 2.0
+
             for i, j in product(range(n_classes), range(n_classes)):
                 color = cmap_max if cm[i, j] < thresh else cmap_min
-                self.text_[i, j] = ax.text(j, i,
-                                           format(cm[i, j], values_format),
-                                           ha="center", va="center",
-                                           color=color)
+
+                fmt = values_format
+                if values_format is None:
+                    fmt = '.2g'
+                    if cm.dtype.kind != 'f':
+                        fmt = 'd'
+                        if len(format(cm[i, j], '.2g')) < len(
+                                                        format(cm[i, j], 'd')):
+                            fmt = '.2g'
+
+                self.text_[i, j] = ax.text(
```

Comment:
Sorry I was not more clear. The current PR may call `format` three times. I am trying to push for calling it at most twice:

```py
if values_format is None:
    text_cm = format(cm[i, j], '.2g')
    if cm.dtype.kind != 'f':
        text_d = format(cm[i, j], 'd')
        if len(text_d) < len(text_cm):
            text_cm = text_d
else:
    text_cm = format(cm[i, j], values_format)

self.text_[i, j] = ax.text(j, i, text_cm, ...)
```